### ISSUE TITLE
Documentation on what ASDF format versions mean

### ISSUE THREAD

**Issue by embray** [spacetelescope/asdf-standard#90]
In the introduction to the file layout, the documentation for the ASDF [header](http://asdf-standard.readthedocs.org/en/latest/file_layout.html#header) explicitly breaks the ASDF version into a "major", "minor", and "micro" version.  But we don't really discuss how versions will be used for the format.  Since there is currently only one version it may yet be premature to say much about it, so I'm putting this here as a reminder.

In particular I think we need to think about how the format version relates to the ASDF standard and to ASDF software.  Of course, there's also room for flexibility and change in how we interpret versions (to that end  we may even want to think ahead an add an optional "epoch" field to the version specification).

Also, I don't think the version of the _standard_, as a document, and the version of the format are always necessarily identical either.  For example, we want to be able to release updates to the standard that patch typos or include clarifications or fill in missing information, but that do not actually change _anything_ about the format or even the schemas.  Though in the interest of keeping things sane I think we still want to keep the standard version in sync with the format version (for example one type of bug fix might be to patch a missing or incorrect rule in one of the schemas, which amounts to an update both in the format and the standard).  To that end I think the version of the standard may also include `.post<N>` tags for minor text updates that don't change the format.

How the format version relates to software versions is trickier and may benefit from experience.  I definitely don't think the two should be tied in any way--the version of an ASDF library should not in any way be tied to what ASDF versions it supports (although that should be well documented).  More importantly will be determining when updates to the ASDF _format_ require updated software.  My feeling is that ASDF files written with newer software should at least be readable by older software the supports the same _major_ version.  In other words, an increase in _minor_ version may have some benefits, but should still be backwards compatible with older software.  Only increases in _major_ versions may introduce backwards incompatibility with older software.  I think that's a good, conservative guideline.  Though I'm not 100% sure how it will be in practice...

**Comment by mdboom** [comment]
I agree with this -- particularly that the library and the specification versions should not be tied to one another.  I think it's not too premature to develop at least some aspirational language about this.

Is this a fair summary for the version of the standard itself...?

- major: Introduces a backward incompatible change.
- minor: Introduces a substantial change which does not break backward compatibility.  i.e. libraries that don't understand the new changes can still read the file, but may not be able to act on the new features, though all old functionality would still work.
- micro: A minor change or clarification to the text of the standard that doesn't impact the file format itself.  (This means that the micro number in as ASDF file should basically always be ignored by a library).

Another thing I've wondered about is how versions of tags (i.e. individual schema) relate to the version of the ASDF specification itself.  Initially I thought they would move in lock step, but now I'm not so sure.  Let's say we release what we have now as 1.0.  Then we release a 1.1, whose only changes are adding a bunch of new transformations.  An older library that doesn't know or care about 1.1 should still be able to handle the ndarray stuff without any problems.  If we update the version number on ndarray to 1.1 as part of the 1.1 release of the specification, there's no way the old library could be certain that it was supporting everything 100% correctly for ndarray objects and could reasonably display a warning -- but in this case the warning would actually be unwarranted.  If we keep the version number of ndarray tags at 1.0, the old library could be confident it was still doing the right thing.  You could also imagine "ndarray 2.0", which adds, say, tiling support in a backward incompatible way.  This would probably necessitate a major version bump in ASDF as well, but ASDF 2.0 could allow both "ndarray 1.0" and "ndarray 2.0" simultaneously.  It's certainly messier to not have all the version numbers match, but it would be easy enough to have a table of the latest version of each schema included with each version of the standard.

**Comment by mdboom** [comment]
One other point: if we do decide that schemas are individually versioned, we should probably change their paths.  Currently, they look like:

```
tag:stsci.edu:asdf/0.1.0/core/ndarray
```

But should probably change to:

```
tag:stsci.edu:asdf/core/ndarray-0.1.0
```

so as not to create a strange heirarchy.

**Comment by embray** [comment]
> A minor change or clarification to the text of the standard that doesn't impact the file format itself.

Originally I was going to saw that the file format *can* be impacted if there are clarifications or minor updates to any of the core schemas.  But if the schemas will be individually versioned then that would just mean an update to the schema version, and that has no impact of course on the ASDF format itself.

 I agree with the tag path change, if the schemas are to be individually versioned.

**Comment by drdavella** [comment]
Closing since this has already been addressed (in #91).
